src/openms/include/OpenMS/FILTERING/CALIBRATION/PrecursorCorrection.h

```diff
@@ -0,0 +1,163 @@
+  writeHist: Write output .csv for validation purposes (corrected, uncorrected).
+  correctToNearestMS1Peak: Correct to the peak in closest proximity in a certrain mass range.
+  correctToHighestIntensityMS1Peak: Correct to the peak with the highest intensity in a certain mass range.
+  correctToNearestFeature: Use feature information to reannotate a precursor (e.g. falsely assigned to non mono-isotopic trace).
+  */
+class OPENMS_DLLAPI PrecursorCorrection
+{
+    public:
+    static const std::string csv_header;
+
+    /**
+    @brief Extract precursors and associated information (precursor retention time and precursor scan index)
+    */
+    static void getPrecursors(const MSExperiment & exp,
+                              std::vector<Precursor> & precursors,
+                              std::vector<double> & precursors_rt,
+                              std::vector<Size> & precursor_scan_index);
+
+
+    /**
+    @brief Writer can be used in association with correctToNearestMS1Peak or correctToHighestIntensityMS1Peak
+
+    Format:
+    RT	    uncorrectedMZ	correctedMZ	deltaMZ
+    100.1	509.9999	    510	         0.0001
+    180.9	610.0001	    610	        -0.0001
+    183.92	611.0035	    611.0033	  -0.0002
+    */
+    static void writeHist(const String& out_csv,
+                          const std::vector<double> & delta_mzs,
+                          const std::vector<double> & mzs,
+                          const std::vector<double> & rts);
+    /**
+    @brief Selection of the peak in closest proximity as corrected precursor mass in a given mass range (e.g. precursor mass +/- 0.2 Da)
+
+    For each MS2 spectrum the corresponding MS1 spectrum is determined by using the rt information of the precursor.
+    In the MS1, the peak closest to the uncorrected precursor m/z is selected and used as corrected precursor m/z.
+    */
+    static std::set<Size> correctToNearestMS1Peak(MSExperiment & exp,
+                                                  double mz_tolerance,
+                                                  bool ppm,
+                                                  std::vector<double> & delta_mzs,
+                                                  std::vector<double> & mzs,
+                                                  std::vector<double> & rts);
+
+     /**
+     @brief Selection of the peak with the highest intensity as corrected precursor mass in a given mass range (e.g. precursor mass +/- 0.2 Da)
+
+     For each MS2 spectrum the corresponding MS1 spectrum is determined by using the rt information of the precursor.
+     In the MS1, the peak with the highest intensity in a given mass range to the uncorrected precursor m/z is selected and used as corrected precursor m/z.
+     */
+     static std::set<Size> correctToHighestIntensityMS1Peak(MSExperiment & exp,
+                                                           double mz_tolerance,
```

Comment:
how come no ppm tolerance here?

Reply:
Can add the ppm tolerance (in next PR?).